Client/Assets/Favicons.js

```diff
@@ -0,0 +1,46 @@
+(function() {
+ // These integers should be kept in sync with the IconType raw-values
+ var ICON = 0;
+ var APPLE = 1;
+ var APPLE_PRECOMPOSED = 2;
+ var GUESS = 3;
+
+ var Favicons = {
+  selectors: { "link[rel~='icon']": ICON,
+              "link[rel='apple-touch-icon']": APPLE,
+              "link[rel='apple-touch-icon-precomposed']": APPLE_PRECOMPOSED },
+
+ getAll: function() {
+    var res = {}
+    var foundIcons = false
+
+    for (selector in this.selectors) {
+      var icons = document.querySelectorAll(selector)
+      for (var i = 0; i < icons.length; i++) {
+        var href = icons[i].href;
+        res[href] = this.selectors[selector];
+        if (!foundIcons && this.selectors[selector] == ICON) {
+          foundIcons = true
+        }
+      }
+    }
+
+    // If we didn't find anything in the page, look to see if a favicon.ico file exists for the domain
+    if (!foundIcons) {
```

Comment:
This is probably a stylistic thing a bit, but I wanted to make sure we always returned an object (not undefined or null). It also means we don't need an if (!res) res = {} line in the loop, but its mostly the former that seems like a pain.